### ISSUE TITLE
dynamic window title setting non functional when TERM=alacritty

### ISSUE THREAD

**Issue by jmesmon** [jwilm/alacritty#1636]
This is on Linux X11.

In xterm, urxvt, and kitty, the title while bash is running is `username@hostname:bash-cwd` (so commonly: `me@my_machine:~` and similar).

Alacritty only shows `Alacritty` as the window title. Window title ~~does change (as expected) when vim or another cmdline application is started that sets it, but~~ in bash: no change from `Alacritty`.

**Comment by jmesmon** [comment]
Actually, looks like no title setting works, and it's related to a configuration change I made:

When `env.TERM` is left unset (and thus is set to `alacritty`), none of the window titles are set.

`infocmp alacritty` returns some terminfo (so it appears the terminfo is installed).

**Comment by chrisduerr** [comment]
Does everything works correctly when setting `TERM: xterm-256color`?

If it works with the xterm terminfo, it's probably either a missing capability in our terminfo file (which seems unlikely since we copied xterm-256color), or bash is checking for the terminal emulator before setting the title.

Also, please make sure that `dynamic_title` is set to `true` in the configuration file.

Note that I manually checked the source code of `bash` and wasn't able to find any mention of `kitty` or `rxvt`, so I'm not sure what exactly is going on. Are you using the `--title` option to override Alacritty's window title?

**Comment by jmesmon** [comment]
Yes, everything appears to work (wrt dynamic titles) when `TERM: xterm-256color` is used.

No, I am not using alacritty's `--title` argument.

**Comment by jwilm** [comment]
@dnkl any idea off the top of your head?

**Comment by dnkl** [comment]
Well, `echo -e "\033]0;hello\007"` obviously works, and `infocmp alacritty xterm-256color` shows no related differences. So, most likely someone somewhere has a list of terminfo names. And alacritty isn't on it.

@jmesmon: there isn't much context in your description; my guess is your window title is supposed to be set by `PS1`? It's usually that, or `PROMPT_COMMAND`.

Next step is to figure out how and where it is set. If you're not sure about all this, start with posting the values of both of them:

```sh
echo $PS1
echo $PROMPT_COMMAND
```

(for the record, I use a custom window title setter in zsh which works just fine with alacritty).

**Comment by dnkl** [comment]
Found a random (unknown which version) "default" Ubuntu bashrc, where they do this:

```sh
# If this is an xterm set the title to user@host:dir
case "$TERM" in
xterm*|rxvt*)
    PS1="\[\e]0;${debian_chroot:+($debian_chroot)}\u@\h: \w\a\]$PS1"
    ;;
*)
    ;;
esac
```

@jmesmon: I'm guessing you have something like this in your (or the system's) bashrc.

**Comment by jmesmon** [comment]
Yep, that's the problem. This is on archlinux. Here's it's /etc/bash.bashrc:
```
$ cat /etc/bash.bashrc
#
# /etc/bash.bashrc
#

# If not running interactively, don't do anything
[[ $- != *i* ]] && return

[[ $DISPLAY ]] && shopt -s checkwinsize

PS1='[\u@\h \W]\$ '

case ${TERM} in
  xterm*|rxvt*|Eterm|aterm|kterm|gnome*)
    PROMPT_COMMAND=${PROMPT_COMMAND:+$PROMPT_COMMAND; }'printf "\033]0;%s@%s:%s\007" "${USER}" "${HOSTNAME%%.*}" "${PWD/#$HOME/\~}"'

    ;;
  screen*)
    PROMPT_COMMAND=${PROMPT_COMMAND:+$PROMPT_COMMAND; }'printf "\033_%s@%s:%s\033\\" "${USER}" "${HOSTNAME%%.*}" "${PWD/#$HOME/\~}"'
    ;;
esac

[ -r /usr/share/bash-completion/bash_completion   ] && . /usr/share/bash-completion/bash_completion
```

**Comment by jmesmon** [comment]
That said, `vim` doesn't update the title. Probably a similar thing.

I did investigate why kitty gets functional dynamic titles: it uses `TERM=xterm-kitty`.

**Comment by jmesmon** [comment]
`vim` appears to have a similar thing, but hard coded in it's source: `src/os_unix.c`:
```
    int
vim_is_xterm(char_u *name)
{
    if (name == NULL)
	return FALSE;
    return (STRNICMP(name, "xterm", 5) == 0
		|| STRNICMP(name, "nxterm", 6) == 0
		|| STRNICMP(name, "kterm", 5) == 0
		|| STRNICMP(name, "mlterm", 6) == 0
		|| STRNICMP(name, "rxvt", 4) == 0
		|| STRNICMP(name, "screen.xterm", 12) == 0
		|| STRCMP(name, "builtin_xterm") == 0);
}
```

**Comment by RunningDroid** [comment]
> If it works with the xterm terminfo, it's probably either a missing capability in our terminfo file (which seems unlikely since we copied xterm-256color)

Before alacritty had it's own terminfo I made a simple [one](https://gist.github.com/RunningDroid/61d75a5f13c448ed4514e7a0cf189411), in the process I learned that xterm-256color's terminfo doesn't list statusline capabilities despite xterm-256color having them, which led to people hard-coding statusline codes for xterm-256color instead of getting them from terminfo.

In short, alacritty.info is missing `hs` (bool), `dsl=\E]2;\007`, `fsl=^G`, and `tsl=\E]2;`.

**Comment by dnkl** [comment]
@RunningDroid: thanks, and nice catch!

For the record, a terminal emulator _could_ choose to implement the statusline with a different mechanism than the window title (for example, the standard talks about fsl/tsl moving the cursor, so it _could_ be a part of the actual console).

But, since most emulators seems to use the window title, I guess that makes the most sense for Alacritty too.

I've verified that adding `hs, dsl=\E]2;\007, fsl=^G, tsl=\E]2;,` to `alacritty.info` fixes the window title / statusline in vim.

@RunningDroid: feel free to submit a pull request, if you're interested. Otherwise @chrisduerr or I can do it.
